src/touch_save_app.c

```diff
@@ -0,0 +1,520 @@
+    .charBase = GX_BG_CHARBASE_0x04000,
+    .bgExtPltt = GX_BG_EXTPLTT_01,
+    .priority = 1,
+    .areaOver = GX_BG_AREAOVER_XLU,
+    .dummy = 0,
+    .mosaic = FALSE,
+};
+
+static const BgTemplate ov30_0225DC48 = {
+    .x = 0,
+    .y = 0,
+    .bufferSize = 0x800,
+    .baseTile = 0,
+    .size = GF_BG_SCR_SIZE_256x256,
+    .colorMode = GX_BG_COLORMODE_16,
+    .screenBase = GX_BG_SCRBASE_0x7000,
+    .charBase = GX_BG_CHARBASE_0x00000,
+    .bgExtPltt = GX_BG_EXTPLTT_01,
+    .priority = 0,
+    .areaOver = GX_BG_AREAOVER_XLU,
+    .dummy = 0,
+    .mosaic = FALSE,
+};
+
+static const BgTemplate ov30_0225DC2C = {
+    .x = 0,
+    .y = 0,
+    .bufferSize = 0x800,
+    .baseTile = 0,
+    .size = GF_BG_SCR_SIZE_256x256,
+    .colorMode = GX_BG_COLORMODE_16,
+    .screenBase = GX_BG_SCRBASE_0x6800,
+    .charBase = GX_BG_CHARBASE_0x00000,
+    .bgExtPltt = GX_BG_EXTPLTT_01,
+    .priority = 2,
+    .areaOver = GX_BG_AREAOVER_XLU,
+    .dummy = 0,
+    .mosaic = FALSE,
+};
+
+typedef BOOL (*TouchSaveAppStateFunc)(TouchSaveAppData*);
+static TouchSaveAppStateFunc sTouchSaveAppStateFuncs[TOUCHSAVEAPP_STATE_COUNT] = {
+	TouchSaveApp_DisplaySaveInformation,
+	TouchSaveApp_AskForSave,
+	TouchSaveApp_WaitForTextPrinter,
+	TouchSaveApp_GetSaveConfirmation,
+	TouchSaveApp_HandleSaveConfirmation,
+	TouchSaveApp_PrintOverwriteMessage,
+	TouchSaveApp_GetOverwriteConfirmation,
+	TouchSaveApp_HandleOverwriteConfirmation,
+	TouchSaveApp_PrintNotMySaveError,
+	TouchSaveApp_SaveFailed,
+	TouchSaveApp_PrintSavingMessage,
+	TouchSaveApp_SetupWaitingIcon,
+	TouchSaveApp_SaveGame,
+	TouchSaveApp_PrintSavedMessage,
+	TouchSaveApp_SaveSucceeded,
+	TouchSaveApp_CloseApp,
+};
+
+SysTask *ov30_0225D520(BgConfig *bgConfig, void* a1, FieldSystem *fieldSystem, void* a3) {
+    CreateHeap(HEAP_ID_3, HEAP_ID_8, 0x18000);
+
+    GXS_SetGraphicsMode(GX_BGMODE_0);
+    GX_SetBankForSubBG(GX_VRAM_SUB_BG_32_H);
+    GX_SetBankForSubOBJ(GX_VRAM_SUB_OBJ_16_I);
+    // FIXME: Make an inline function for this.
+    reg_GXS_DB_DISPCNT = (reg_GXS_DB_DISPCNT & ~(REG_GXS_DB_DISPCNT_EXOBJ_CH_MASK | REG_GXS_DB_DISPCNT_OBJMAP_CH_MASK)) | REG_GXS_DB_DISPCNT_OBJMAP_CH_MASK;
+
+    InitBgFromTemplate(bgConfig, 4, &ov30_0225DC2C, GF_BG_TYPE_TEXT);
+    InitBgFromTemplate(bgConfig, 5, &ov30_0225DC48, GF_BG_TYPE_TEXT);
+    InitBgFromTemplate(bgConfig, 6, &ov30_0225DC64, GF_BG_TYPE_TEXT);
+    BG_ClearCharDataRange(6, 0x20, 0, HEAP_ID_4);
+    BgClearTilemapBufferAndCommit(bgConfig, 6);
+    BgClearTilemapBufferAndCommit(bgConfig, 5);
+
+    GXS_SetVisibleWnd(0);
+    GX_EngineBToggleLayers(1, GX_LAYER_TOGGLE_OFF);
+    GX_EngineBToggleLayers(2, GX_LAYER_TOGGLE_OFF);
+    GX_EngineBToggleLayers(4, GX_LAYER_TOGGLE_OFF);
+    GX_EngineBToggleLayers(16, GX_LAYER_TOGGLE_OFF);
+
+    SysTask *task = sub_02007200((SysTaskFunc)ov30_0225D700, sizeof(TouchSaveAppData), 10, HEAP_ID_8);
+    TouchSaveAppData *data = sub_0201F988(task);
+    data->task = task;
+    data->unk0 = 0;
```

Comment:
NULL instead of 0

Reply:
I've retyped this to a u32 since it may not be a pointer (it's completely unused)